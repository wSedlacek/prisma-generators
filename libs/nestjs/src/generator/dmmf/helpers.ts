// TODO: support not string args
export const modelAttributeRegex = /(@@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export const parseDocumentationAttributes = (
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: 'model' | 'field'
) => {
  const attributeRegex =
    expectedAttributeKind === 'model'
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return undefined;
  }

  return attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
};
