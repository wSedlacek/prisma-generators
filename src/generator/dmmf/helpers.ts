// TODO: support not string args
export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export function parseDocumentationAttributes(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
) {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return;
  }
  const attributeArgs = attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
  return attributeArgs;
}
