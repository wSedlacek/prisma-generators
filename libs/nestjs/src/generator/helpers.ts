import { DmmfDocument } from './dmmf/dmmf-document';
import { fieldAttributeRegex, modelAttributeRegex } from './dmmf/helpers';
import { DMMF } from './dmmf/types';

export const noop = () => {};

export const getFieldTSType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  isInputType: boolean,
  modelName?: string,
  typeName?: string
) => {
  let TSType: string;
  if (typeInfo.kind === 'scalar') {
    TSType = mapScalarToTSType(typeInfo.type, isInputType);
  } else if (typeInfo.kind === 'object') {
    TSType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else if (typeInfo.kind === 'enum') {
    TSType = `keyof typeof ${typeInfo.type}`;
  } else {
    throw new Error(`Unsupported field type kind: ${typeInfo.kind}`);
  }
  if (typeInfo.isList) {
    if (TSType.includes(' ')) {
      TSType = `Array<${TSType}>`;
    } else {
      TSType += '[]';
    }
  }
  if (!typeInfo.isRequired) {
    // FIXME: use properly null for output and undefined for input
    // TSType += ' | undefined';
    TSType += ' | null | undefined';
  }

  return TSType;
};

export const mapScalarToTSType = (scalar: string, isInputType: boolean) => {
  switch (scalar) {
    case 'ID':
    case 'UUID': {
      return 'string';
    }
    case 'String': {
      return 'string';
    }
    case 'Boolean': {
      return 'boolean';
    }
    case 'DateTime': {
      return 'Date';
    }
    case 'Int':
    case 'Float': {
      return 'number';
    }
    case 'Json':
      return isInputType ? 'InputJsonValue' : 'JsonValue';
    default:
      throw new Error(`Unrecognized scalar type: ${scalar}`);
  }
};

export const getGraphQLType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  modelName?: string,
  typeName?: string
) => {
  let GraphQLType: string;
  if (typeInfo.kind === 'scalar') {
    GraphQLType = mapScalarToType(typeInfo.type);
  } else if (typeInfo.kind === 'object') {
    GraphQLType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else {
    GraphQLType = typeInfo.type;
  }
  if (typeInfo.isList) {
    GraphQLType = `[${GraphQLType}]`;
  }

  return GraphQLType;
};

export const mapScalarToType = (scalar: string) => {
  switch (scalar) {
    case 'DateTime': {
      return 'Date';
    }
    // TODO: use proper uuid graphql scalar
    case 'UUID': {
      return 'String';
    }
    case 'Boolean':
    case 'String': {
      return scalar;
    }
    case 'ID':
    case 'Int':
    case 'Float': {
      return `${scalar}`;
    }
    case 'Json': {
      return `GraphQLJSON`;
    }
    default: {
      throw new Error(`Unrecognized scalar type: ${scalar}`);
    }
  }
};

export const camelCase = (str: string) => {
  return str[0].toLowerCase() + str.slice(1);
};

export const pascalCase = (str: string): string => {
  return str[0].toUpperCase() + str.slice(1);
};

const getInputKeywordPhrasePosition = (inputTypeName: string) => {
  const inputParseResult = [
    'Create',
    'OrderBy',
    'Update',
    'Upsert',
    'ScalarWhere',
    'Where',
    'ListRelationFilter',
    'RelationFilter',
    'Filter',
  ]
    .map((inputKeyword) => inputTypeName.search(inputKeyword))
    .filter((position) => position >= 0);

  if (inputParseResult.length === 0) {
    return undefined;
  }

  return inputParseResult[0];
};

export const getModelNameFromInputType = (inputTypeName: string) => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(inputTypeName);
  if (!keywordPhrasePosition) {
    return undefined;
  }

  return inputTypeName.slice(0, keywordPhrasePosition);
};

export const getInputTypeName = (
  originalInputName: string,
  dmmfDocument: DmmfDocument
): string => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(
    originalInputName
  );
  if (!keywordPhrasePosition) {
    return originalInputName;
  }

  const modelName = originalInputName.slice(0, keywordPhrasePosition);
  const typeNameRest = originalInputName.slice(keywordPhrasePosition);
  const modelTypeName = dmmfDocument.getModelTypeName(modelName);
  if (!modelTypeName) {
    return originalInputName;
  }

  return `${modelTypeName}${typeNameRest}`;
};

export const cleanDocsString = (
  documentation: string | undefined
): string | undefined => {
  if (!documentation) {
    return undefined;
  }
  let cleanedDocs = documentation;
  cleanedDocs = cleanedDocs.replace(modelAttributeRegex, '');
  cleanedDocs = cleanedDocs.replace(fieldAttributeRegex, '');
  cleanedDocs = cleanedDocs.split('"').join('\\"');
  cleanedDocs = cleanedDocs.split('\r').join('');
  cleanedDocs = cleanedDocs.split('\\r').join('');
  cleanedDocs = cleanedDocs.split('\n').join('');
  cleanedDocs = cleanedDocs.split('\\n').join('');

  return cleanedDocs;
};

export const toUnixPath = (maybeWindowsPath: string) => {
  return maybeWindowsPath.split('\\').join('/');
};
