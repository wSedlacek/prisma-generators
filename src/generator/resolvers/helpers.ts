import { getFieldTSType, getGraphQLType } from '../helpers';
import { DmmfDocument } from '../dmmf/dmmf-document';
import { DMMF } from '../dmmf/types';

export const generateCrudResolverClassMethodDeclaration = (
  action: DMMF.Action,
  typeName: string,
  method: DMMF.SchemaField,
  argsTypeName: string | undefined,
  collectionName: string,
  dmmfDocument: DmmfDocument,
  mapping: DMMF.Mapping
) => {
  const returnTSType = getFieldTSType(
    method.outputType,
    dmmfDocument,
    false,
    mapping.model,
    typeName
  );

  return {
    name: action.name,
    isAsync: true,
    returnType: `Promise<${returnTSType}>`,
    decorators: [
      {
        name: `${action.operation}`,
        arguments: [
          `() => ${getGraphQLType(
            method.outputType,
            dmmfDocument,
            mapping.model,
            typeName
          )}`,
          `{
            ${[
              `nullable: ${!method.outputType.isRequired}`,
              `description: undefined`,
            ].join(',\n')}
          }`,
        ],
      },
    ],
    parameters: [
      {
        name: 'ctx',
        // TODO: import custom `ContextType`
        type: 'any',
        decorators: [{ name: 'Context', arguments: [] }],
      },
      ...(action.kind === 'aggregate'
        ? [
            {
              name: 'info',
              type: 'GraphQLResolveInfo',
              decorators: [{ name: 'Info', arguments: [] }],
            },
          ]
        : []),
      ...(!argsTypeName
        ? []
        : [
            {
              name: 'args',
              type: argsTypeName,
              decorators: [{ name: 'Args', arguments: [] }],
            },
          ]),
    ],
    statements:
      action.kind === 'aggregate'
        ? [
            `const transformFields = (fields: Record<string, any>): Record<string, any> => {
              return Object.fromEntries(
                Object.entries(fields)
                  .filter(([key, value]) => !key.startsWith("_"))
                  .map<[string, any]>(([key, value]) => {
                    if (Object.keys(value).length === 0) {
                      return [key, true];
                    }
                    return [key, transformFields(value)];
                  }),
              );
            }`,
            `return ctx.prisma.${collectionName}.${action.kind}({
              ...args,
              ...transformFields(graphqlFields(info as any)),
            });`,
          ]
        : [
            `return plainToClass(${returnTSType
              .replace(/\[|\]/g, '')
              .replace(
                '| undefined',
                ''
              )}, await ctx.prisma.${collectionName}.${action.kind}(${
              argsTypeName ? 'args' : ''
            }) as ${getGraphQLType(
              method.outputType,
              dmmfDocument,
              mapping.model,
              typeName
            )});`,
          ],
  };
};
