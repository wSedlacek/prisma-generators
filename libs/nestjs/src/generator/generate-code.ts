import { DMMF as PrismaDMMF } from '@prisma/client/runtime/dmmf-types';
import path from 'path';
import { CompilerOptions, ModuleKind, Project, ScriptTarget } from 'ts-morph';

import { isDefined } from '../utils';
import { generateArgsTypeClassFromArgs } from './args-class';
import {
  argsFolderName,
  crudResolversFolderName,
  enumsFolderName,
  inputsFolderName,
  modelsFolderName,
  outputsFolderName,
  relationsResolversFolderName,
  resolversFolderName,
} from './config';
import { DmmfDocument } from './dmmf/dmmf-document';
import { DMMF } from './dmmf/types';
import { generateEnumFromDef } from './enum';
import { noop } from './helpers';
import {
  generateArgsBarrelFile,
  generateEnumsBarrelFile,
  generateIndexFile,
  generateInputsBarrelFile,
  generateModelsBarrelFile,
  generateOutputsBarrelFile,
  generateResolversBarrelFile,
} from './imports';
import { generateObjectTypeClassFromModel } from './model-type-class';
import { GenerateCodeOptions } from './options';
import { generateCrudResolverClassFromMapping } from './resolvers/full-crud';
import { generateRelationsResolverClassesFromModel } from './resolvers/relations';
import { generateActionResolverClass } from './resolvers/separate-action';
import {
  generateInputTypeClassFromType,
  generateOutputTypeClassFromType,
} from './type-class';

const baseCompilerOptions: CompilerOptions = {
  target: ScriptTarget.ES2019,
  module: ModuleKind.CommonJS,
  emitDecoratorMetadata: true,
  experimentalDecorators: true,
};

export const generateCode = async (
  dmmf: PrismaDMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop
) => {
  const baseDirPath = options.outputDirPath;
  const emitTranspiledCode =
    options.emitTranspiledCode ??
    options.outputDirPath.includes('node_modules');
  const project = new Project({
    compilerOptions: {
      ...baseCompilerOptions,
      ...(emitTranspiledCode ? { declaration: true } : {}),
    },
  });
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map((model) => model.name);

  log('Transforming dmmfDocument...');
  const dmmfDocument = new DmmfDocument(dmmf, options);

  log('Generating enums...');
  const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
    (enumDef) => enumDef.typeName
  );
  dmmfDocument.datamodel.enums.forEach((enumDef) =>
    generateEnumFromDef(project, baseDirPath, enumDef)
  );
  dmmfDocument.schema.enums
    // skip enums from datamodel
    .filter((enumDef) => !datamodelEnumNames.includes(enumDef.typeName))
    .forEach((enumDef) => generateEnumFromDef(project, baseDirPath, enumDef));
  const emittedEnumNames = [
    ...new Set([
      ...dmmfDocument.schema.enums.map((it) => it.typeName),
      ...dmmfDocument.datamodel.enums.map((it) => it.typeName),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);

  log('Generating models...');
  dmmfDocument.datamodel.models.forEach((model) =>
    generateObjectTypeClassFromModel(project, baseDirPath, model, dmmfDocument)
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmfDocument.datamodel.models.map((it) => it.typeName)
  );

  log('Generating output types...');
  const rootTypes = dmmfDocument.schema.outputTypes.filter((type) =>
    ['Query', 'Mutation'].includes(type.name)
  );
  const outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
    // skip generating models and root resolvers
    (type) => !modelNames.includes(type.name) && !rootTypes.includes(type)
  );
  const outputTypesFieldsArgsToGenerate = outputTypesToGenerate
    .map((it) => it.fields)
    .reduce((a, b) => a.concat(b), [])
    .filter((it): it is Required<DMMF.OutputSchemaField> => !!it.argsTypeName);
  outputTypesToGenerate.forEach((type) =>
    generateOutputTypeClassFromType(
      project,
      resolversDirPath,
      type,
      dmmfDocument
    )
  );
  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map((it) => it.typeName),
    outputTypesFieldsArgsToGenerate.length > 0
  );

  if (outputTypesFieldsArgsToGenerate.length > 0) {
    log('Generating output types args...');
    outputTypesFieldsArgsToGenerate.forEach((field) => {
      generateArgsTypeClassFromArgs(
        project,
        path.resolve(resolversDirPath, outputsFolderName),
        field.args,
        field.argsTypeName,
        dmmfDocument,
        2
      );
    });

    const outputsArgsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        argsFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateArgsBarrelFile(
      outputsArgsBarrelExportSourceFile,
      outputTypesFieldsArgsToGenerate
        .map((it) => it.argsTypeName)
        .filter(isDefined)
    );
  }

  log('Generating input types...');

  dmmfDocument.schema.inputTypes.forEach((type) =>
    generateInputTypeClassFromType(
      project,
      resolversDirPath,
      type,
      dmmfDocument
    )
  );

  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmfDocument.schema.inputTypes.map((it) => it.typeName)
  );

  if (dmmfDocument.relationModels.length > 0) {
    log('Generating relation resolvers...');

    dmmfDocument.relationModels.forEach((relationModel) =>
      generateRelationsResolverClassesFromModel(
        project,
        baseDirPath,
        dmmfDocument,
        relationModel
      )
    );

    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateResolversBarrelFile(
      'relations',
      relationResolversBarrelExportSourceFile,
      dmmfDocument.relationModels.map((relationModel) => ({
        resolverName: relationModel.resolverName,
        modelName: relationModel.model.typeName,
        hasSomeArgs: relationModel.relationFields.some(
          (field) => field.argsTypeName !== undefined
        ),
      }))
    );

    log('Generating relation resolver args...');
    dmmfDocument.relationModels.forEach((relationModelData) => {
      const resolverDirPath = path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        relationModelData.model.typeName
      );

      relationModelData.relationFields
        .filter((field) => field.argsTypeName)
        .forEach((field) => {
          generateArgsTypeClassFromArgs(
            project,
            resolverDirPath,
            field.outputTypeField.args,
            field.argsTypeName as string,
            dmmfDocument
          );
        });

      const argTypeNames = relationModelData.relationFields
        .filter((it) => it.argsTypeName !== undefined)
        .map((it) => it.argsTypeName)
        .filter(isDefined);

      if (argTypeNames.length) {
        const barrelExportSourceFile = project.createSourceFile(
          path.resolve(resolverDirPath, argsFolderName, 'index.ts'),
          undefined,
          { overwrite: true }
        );
        generateArgsBarrelFile(barrelExportSourceFile, argTypeNames);
      }
    });
  }

  log('Generating crud resolvers...');
  dmmfDocument.mappings.forEach((mapping) => {
    const model = dmmfDocument.datamodel.models.find(
      (it) => it.name === mapping.model
    ) as DMMF.Model;
    generateCrudResolverClassFromMapping(
      project,
      baseDirPath,
      mapping,
      model,
      dmmfDocument
    );
    mapping.actions.forEach((action) => {
      const actionModel = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;
      generateActionResolverClass(
        project,
        baseDirPath,
        actionModel,
        action,
        mapping,
        dmmfDocument
      );
    });
  });
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateResolversBarrelFile(
    'crud',
    crudResolversBarrelExportSourceFile,
    dmmfDocument.mappings.map((mapping) => {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;

      return {
        modelName: model.typeName,
        resolverName: mapping.resolverName,
        actionResolverNames: mapping.actions.map((it) => it.actionResolverName),
        hasSomeArgs: mapping.actions.some(
          (action) => action.argsTypeName !== undefined
        ),
      };
    })
  );

  log('Generating crud resolvers args...');
  dmmfDocument.mappings.forEach((mapping) => {
    const actionsWithArgs = mapping.actions.filter(
      (it) => it.argsTypeName !== undefined
    );

    if (actionsWithArgs.length) {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;
      const resolverDirPath = path.resolve(
        baseDirPath,
        resolversFolderName,
        crudResolversFolderName,
        model.typeName
      );
      actionsWithArgs.forEach((action) => {
        generateArgsTypeClassFromArgs(
          project,
          resolverDirPath,
          action.method.args,
          action.argsTypeName as string,
          dmmfDocument
        );
      });
      const barrelExportSourceFile = project.createSourceFile(
        path.resolve(resolverDirPath, argsFolderName, 'index.ts'),
        undefined,
        { overwrite: true }
      );
      generateArgsBarrelFile(
        barrelExportSourceFile,
        actionsWithArgs.map((it) => it.argsTypeName).filter(isDefined)
      );
    }
  });

  log('Generating index file');
  const indexSourceFile = project.createSourceFile(
    `${baseDirPath}/index.ts`,
    undefined,
    { overwrite: true }
  );
  generateIndexFile(indexSourceFile, dmmfDocument.relationModels.length > 0);

  log('Emitting generated code files');
  if (emitTranspiledCode) {
    await project.emit();
  } else {
    for (const file of project.getSourceFiles()) {
      file.formatText({ indentSize: 2 });
    }
    await project.save();
  }
};
