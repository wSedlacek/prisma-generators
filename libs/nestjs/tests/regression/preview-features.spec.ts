import { promises as fs } from 'fs';

import { generateArtifactsDirPath } from '../helpers/artifacts-dir';
import { generateCodeFromSchema } from '../helpers/generate-code';
import { prisma } from '../helpers/prisma-template';
import {
  createReadGeneratedFile,
  ReadGeneratedFile,
} from '../helpers/read-file';

describe('preview features', () => {
  let outputDirPath: string;
  let readGeneratedFile: ReadGeneratedFile;

  beforeEach(async () => {
    outputDirPath = generateArtifactsDirPath('preview-features');
    await fs.mkdir(outputDirPath, { recursive: true });
    readGeneratedFile = createReadGeneratedFile(outputDirPath);
  });

  describe('when `connectOrCreate` is enabled', () => {
    it('should properly generate input type classes for connectOrCreate', async () => {
      const schema = prisma`
        model User {
          id          Int     @id @default(autoincrement())
          name        String
          postsField  Post[]
        }
        model Post {
          id        Int     @id @default(autoincrement())
          title     String
          authorId  Int
          author    User    @relation(fields: [authorId], references: [id])
        }
      `;

      await generateCodeFromSchema(schema, {
        outputDirPath,
        enabledPreviewFeatures: ['connectOrCreate'],
      });
      const userUpdateOneRequiredWithoutPostsFieldInputTSFile = await readGeneratedFile(
        '/resolvers/inputs/UserUpdateOneRequiredWithoutPostsFieldInput.ts'
      );
      const userCreateOrConnectWithoutpostInputTSFile = await readGeneratedFile(
        '/resolvers/inputs/UserCreateOrConnectWithoutPostInput.ts'
      );
      const userCreateOneWithoutPostsFieldInputTSFile = await readGeneratedFile(
        '/resolvers/inputs/UserCreateOneWithoutPostsFieldInput.ts'
      );
      const inputsIndexTSFile = await readGeneratedFile(
        '/resolvers/inputs/index.ts'
      );

      expect(userUpdateOneRequiredWithoutPostsFieldInputTSFile).toMatchSnapshot(
        'UserUpdateOneRequiredWithoutPostsFieldInput'
      );
      expect(userCreateOrConnectWithoutpostInputTSFile).toMatchSnapshot(
        'UserCreateOrConnectWithoutPostInput'
      );
      expect(userCreateOneWithoutPostsFieldInputTSFile).toMatchSnapshot(
        'UserCreateOneWithoutPostsFieldInput'
      );
      expect(inputsIndexTSFile).toMatchSnapshot('inputs index');
    });
  });
});
