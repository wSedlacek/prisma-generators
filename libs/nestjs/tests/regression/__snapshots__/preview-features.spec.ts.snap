// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preview features when \`connectOrCreate\` is enabled should properly generate input type classes for connectOrCreate: UserCreateOneWithoutPostsFieldInput 1`] = `
"import { Field, Float, ID, InputType, Int } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";
import { UserCreateOrConnectWithoutPostInput } from \\"../inputs/UserCreateOrConnectWithoutPostInput\\";
import { UserCreateWithoutPostsFieldInput } from \\"../inputs/UserCreateWithoutPostsFieldInput\\";
import { UserWhereUniqueInput } from \\"../inputs/UserWhereUniqueInput\\";

@InputType({
  isAbstract: true,
  description: undefined,
})
export class UserCreateOneWithoutPostsFieldInput {
  @ClassTransformer__Type(() => UserCreateWithoutPostsFieldInput)
  @Field(() => UserCreateWithoutPostsFieldInput, {
    nullable: true,
    description: undefined
  })
  create?: UserCreateWithoutPostsFieldInput | null | undefined;

  @ClassTransformer__Type(() => UserWhereUniqueInput)
  @Field(() => UserWhereUniqueInput, {
    nullable: true,
    description: undefined
  })
  connect?: UserWhereUniqueInput | null | undefined;

  @ClassTransformer__Type(() => UserCreateOrConnectWithoutPostInput)
  @Field(() => UserCreateOrConnectWithoutPostInput, {
    nullable: true,
    description: undefined
  })
  connectOrCreate?: UserCreateOrConnectWithoutPostInput | null | undefined;
}
"
`;

exports[`preview features when \`connectOrCreate\` is enabled should properly generate input type classes for connectOrCreate: UserCreateOrConnectWithoutPostInput 1`] = `
"import { Field, Float, ID, InputType, Int } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";
import { UserCreateWithoutPostsFieldInput } from \\"../inputs/UserCreateWithoutPostsFieldInput\\";
import { UserWhereUniqueInput } from \\"../inputs/UserWhereUniqueInput\\";

@InputType({
  isAbstract: true,
  description: undefined,
})
export class UserCreateOrConnectWithoutPostInput {
  @ClassTransformer__Type(() => UserWhereUniqueInput)
  @Field(() => UserWhereUniqueInput, {
    nullable: false,
    description: undefined
  })
  where!: UserWhereUniqueInput;

  @ClassTransformer__Type(() => UserCreateWithoutPostsFieldInput)
  @Field(() => UserCreateWithoutPostsFieldInput, {
    nullable: false,
    description: undefined
  })
  create!: UserCreateWithoutPostsFieldInput;
}
"
`;

exports[`preview features when \`connectOrCreate\` is enabled should properly generate input type classes for connectOrCreate: UserUpdateOneRequiredWithoutPostsFieldInput 1`] = `
"import { Field, Float, ID, InputType, Int } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";
import { UserCreateOrConnectWithoutPostInput } from \\"../inputs/UserCreateOrConnectWithoutPostInput\\";
import { UserCreateWithoutPostsFieldInput } from \\"../inputs/UserCreateWithoutPostsFieldInput\\";
import { UserUpdateWithoutPostsFieldDataInput } from \\"../inputs/UserUpdateWithoutPostsFieldDataInput\\";
import { UserUpsertWithoutPostsFieldInput } from \\"../inputs/UserUpsertWithoutPostsFieldInput\\";
import { UserWhereUniqueInput } from \\"../inputs/UserWhereUniqueInput\\";

@InputType({
  isAbstract: true,
  description: undefined,
})
export class UserUpdateOneRequiredWithoutPostsFieldInput {
  @ClassTransformer__Type(() => UserCreateWithoutPostsFieldInput)
  @Field(() => UserCreateWithoutPostsFieldInput, {
    nullable: true,
    description: undefined
  })
  create?: UserCreateWithoutPostsFieldInput | null | undefined;

  @ClassTransformer__Type(() => UserWhereUniqueInput)
  @Field(() => UserWhereUniqueInput, {
    nullable: true,
    description: undefined
  })
  connect?: UserWhereUniqueInput | null | undefined;

  @ClassTransformer__Type(() => UserUpdateWithoutPostsFieldDataInput)
  @Field(() => UserUpdateWithoutPostsFieldDataInput, {
    nullable: true,
    description: undefined
  })
  update?: UserUpdateWithoutPostsFieldDataInput | null | undefined;

  @ClassTransformer__Type(() => UserUpsertWithoutPostsFieldInput)
  @Field(() => UserUpsertWithoutPostsFieldInput, {
    nullable: true,
    description: undefined
  })
  upsert?: UserUpsertWithoutPostsFieldInput | null | undefined;

  @ClassTransformer__Type(() => UserCreateOrConnectWithoutPostInput)
  @Field(() => UserCreateOrConnectWithoutPostInput, {
    nullable: true,
    description: undefined
  })
  connectOrCreate?: UserCreateOrConnectWithoutPostInput | null | undefined;
}
"
`;

exports[`preview features when \`connectOrCreate\` is enabled should properly generate input type classes for connectOrCreate: inputs index 1`] = `
"export { IntFilter } from \\"./IntFilter\\";
export { NestedIntFilter } from \\"./NestedIntFilter\\";
export { NestedStringFilter } from \\"./NestedStringFilter\\";
export { PostCreateInput } from \\"./PostCreateInput\\";
export { PostCreateManyWithoutAuthorInput } from \\"./PostCreateManyWithoutAuthorInput\\";
export { PostCreateOrConnectWithoutUserInput } from \\"./PostCreateOrConnectWithoutUserInput\\";
export { PostCreateWithoutAuthorInput } from \\"./PostCreateWithoutAuthorInput\\";
export { PostListRelationFilter } from \\"./PostListRelationFilter\\";
export { PostOrderByInput } from \\"./PostOrderByInput\\";
export { PostScalarWhereInput } from \\"./PostScalarWhereInput\\";
export { PostUpdateInput } from \\"./PostUpdateInput\\";
export { PostUpdateManyDataInput } from \\"./PostUpdateManyDataInput\\";
export { PostUpdateManyMutationInput } from \\"./PostUpdateManyMutationInput\\";
export { PostUpdateManyWithWhereNestedInput } from \\"./PostUpdateManyWithWhereNestedInput\\";
export { PostUpdateManyWithoutAuthorInput } from \\"./PostUpdateManyWithoutAuthorInput\\";
export { PostUpdateWithWhereUniqueWithoutAuthorInput } from \\"./PostUpdateWithWhereUniqueWithoutAuthorInput\\";
export { PostUpdateWithoutAuthorDataInput } from \\"./PostUpdateWithoutAuthorDataInput\\";
export { PostUpsertWithWhereUniqueWithoutAuthorInput } from \\"./PostUpsertWithWhereUniqueWithoutAuthorInput\\";
export { PostWhereInput } from \\"./PostWhereInput\\";
export { PostWhereUniqueInput } from \\"./PostWhereUniqueInput\\";
export { StringFieldUpdateOperationsInput } from \\"./StringFieldUpdateOperationsInput\\";
export { StringFilter } from \\"./StringFilter\\";
export { UserCreateInput } from \\"./UserCreateInput\\";
export { UserCreateOneWithoutPostsFieldInput } from \\"./UserCreateOneWithoutPostsFieldInput\\";
export { UserCreateOrConnectWithoutPostInput } from \\"./UserCreateOrConnectWithoutPostInput\\";
export { UserCreateWithoutPostsFieldInput } from \\"./UserCreateWithoutPostsFieldInput\\";
export { UserOrderByInput } from \\"./UserOrderByInput\\";
export { UserRelationFilter } from \\"./UserRelationFilter\\";
export { UserUpdateInput } from \\"./UserUpdateInput\\";
export { UserUpdateManyMutationInput } from \\"./UserUpdateManyMutationInput\\";
export { UserUpdateOneRequiredWithoutPostsFieldInput } from \\"./UserUpdateOneRequiredWithoutPostsFieldInput\\";
export { UserUpdateWithoutPostsFieldDataInput } from \\"./UserUpdateWithoutPostsFieldDataInput\\";
export { UserUpsertWithoutPostsFieldInput } from \\"./UserUpsertWithoutPostsFieldInput\\";
export { UserWhereInput } from \\"./UserWhereInput\\";
export { UserWhereUniqueInput } from \\"./UserWhereUniqueInput\\";
"
`;
